File: src/parser.rs
use pest::Parser;
use pest_derive::Parser;
use crate::core::{Expr, Type};

#[derive(Parser)]
#[grammar = "stlc.pest"]
pub struct STLCParser;

pub fn parse(input: &str) -> Result<Expr, String> {
    let pairs = STLCParser::parse(Rule::program, input)
        .map_err(|e| format!("Parse error: {}", e))?;

    for pair in pairs {
        if pair.as_rule() == Rule::program {
            for inner_pair in pair.into_inner() {
                if inner_pair.as_rule() == Rule::expr {
                    return parse_expr(inner_pair);
                }
            }
        }
    }

    Err("No expression found".to_string())
}

fn parse_expr(pair: pest::iterators::Pair<Rule>) -> Result<Expr, String> {
    match pair.as_rule() {
        Rule::expr => {
            let inner = pair.into_inner().next().unwrap();
            parse_expr(inner)
        }
        Rule::lambda => parse_lambda(pair),
        Rule::fix => parse_fix(pair),
        Rule::if_expr => parse_if(pair),
        Rule::comparison => parse_comparison(pair),
        Rule::arith => parse_arith(pair),
        Rule::term => parse_term(pair),
        Rule::factor => parse_factor(pair),
        Rule::atom | Rule::var => {
            let inner = pair.into_inner().next().unwrap();
            parse_expr(inner)
        }
        Rule::ident => Ok(Expr::Var(pair.as_str().to_string())),
        Rule::integer => {
            let val = pair.as_str().parse::<i64>()
                .map_err(|e| format!("Invalid integer: {}", e))?;
            Ok(Expr::Int(val))
        }
        Rule::boolean => {
            let val = pair.as_str() == "true";
            Ok(Expr::Bool(val))
        }
        _ => Err(format!("Unexpected rule: {:?}", pair.as_rule()))
    }
}

fn parse_lambda(pair: pest::iterators::Pair<Rule>) -> Result<Expr, String> {
    let mut inner = pair.into_inner();
    let param = inner.next().unwrap().as_str().to_string();
    let param_type = parse_type(inner.next().unwrap())?;
    let body = parse_expr(inner.next().unwrap())?;

    Ok(Expr::Lam(param, param_type, Box::new(body)))
}

fn parse_fix(pair: pest::iterators::Pair<Rule>) -> Result<Expr, String> {
    let mut inner = pair.into_inner();
    let name = inner.next().unwrap().as_str().to_string();
    let fix_type = parse_type(inner.next().unwrap())?;
    let body = parse_expr(inner.next().unwrap())?;

    Ok(Expr::Fix(name, fix_type, Box::new(body)))
}

fn parse_if(pair: pest::iterators::Pair<Rule>) -> Result<Expr, String> {
    let mut inner = pair.into_inner();
    let cond = parse_expr(inner.next().unwrap())?;
    let then_branch = parse_expr(inner.next().unwrap())?;
    let else_branch = parse_expr(inner.next().unwrap())?;

    Ok(Expr::If(Box::new(cond), Box::new(then_branch), Box::new(else_branch)))
}

fn parse_comparison(pair: pest::iterators::Pair<Rule>) -> Result<Expr, String> {
    let mut inner = pair.into_inner();
    let mut left = parse_expr(inner.next().unwrap())?;

    while let Some(op_pair) = inner.next() {
        if op_pair.as_rule() == Rule::comp_op {
            let op = op_pair.as_str();
            let right = parse_expr(inner.next().unwrap())?;
            left = match op {
                "==" => Expr::Eq(Box::new(left), Box::new(right)),
                "<" => Expr::Lt(Box::new(left), Box::new(right)),
                ">" => Expr::Gt(Box::new(left), Box::new(right)),
                _ => return Err(format!("Unknown comparison operator: {}", op))
            };
        } else {
            left = parse_expr(op_pair)?;
        }
    }

    Ok(left)
}

fn parse_arith(pair: pest::iterators::Pair<Rule>) -> Result<Expr, String> {
    let mut inner = pair.into_inner();
    let mut left = parse_expr(inner.next().unwrap())?;

    while let Some(op_pair) = inner.next() {
        if op_pair.as_rule() == Rule::add_op {
            let op = op_pair.as_str();
            let right = parse_expr(inner.next().unwrap())?;
            left = match op {
                "+" => Expr::Add(Box::new(left), Box::new(right)),
                "-" => Expr::Sub(Box::new(left), Box::new(right)),
                _ => return Err(format!("Unknown arithmetic operator: {}", op))
            };
        } else {
            left = parse_expr(op_pair)?;
        }
    }

    Ok(left)
}

fn parse_term(pair: pest::iterators::Pair<Rule>) -> Result<Expr, String> {
    let mut inner = pair.into_inner();
    let mut left = parse_expr(inner.next().unwrap())?;

    while let Some(op_pair) = inner.next() {
        if op_pair.as_rule() == Rule::mul_op {
            let op = op_pair.as_str();
            let right = parse_expr(inner.next().unwrap())?;
            left = match op {
                "*" => Expr::Mul(Box::new(left), Box::new(right)),
                _ => return Err(format!("Unknown multiplication operator: {}", op))
            };
        } else {
            left = parse_expr(op_pair)?;
        }
    }

    Ok(left)
}

fn parse_factor(pair: pest::iterators::Pair<Rule>) -> Result<Expr, String> {
    let mut inner = pair.into_inner();
    let mut left = parse_expr(inner.next().unwrap())?;

    // Handle function application
    for arg_pair in inner {
        let right = parse_expr(arg_pair)?;
        left = Expr::App(Box::new(left), Box::new(right));
    }

    Ok(left)
}

fn parse_type(pair: pest::iterators::Pair<Rule>) -> Result<Type, String> {
    match pair.as_rule() {
        Rule::type_expr => {
            let inner = pair.into_inner().next().unwrap();
            parse_type(inner)
        }
        Rule::type_arrow => {
            let mut inner = pair.into_inner();
            let mut left = parse_type(inner.next().unwrap())?;

            // Right-associative arrow types
            if let Some(right_pair) = inner.next() {
                let right = parse_type_arrow_rest(inner, right_pair)?;
                left = Type::Arrow(Box::new(left), Box::new(right));
            }

            Ok(left)
        }
        Rule::type_base => {
            match pair.as_str() {
                "Int" => Ok(Type::Int),
                "Bool" => Ok(Type::Bool),
                _ => Err(format!("Unknown type: {}", pair.as_str()))
            }
        }
        _ => Err(format!("Unexpected type rule: {:?}", pair.as_rule()))
    }
}

fn parse_type_arrow_rest(
    mut inner: pest::iterators::Pairs<Rule>,
    first: pest::iterators::Pair<Rule>
) -> Result<Type, String> {
    let mut left = parse_type(first)?;

    if let Some(right_pair) = inner.next() {
        let right = parse_type_arrow_rest(inner, right_pair)?;
        left = Type::Arrow(Box::new(left), Box::new(right));
    }

    Ok(left)
}

File: src/main/scala/Parser.scala
import scala.util.matching.Regex
import scala.util.parsing.combinator.{PackratParsers, RegexParsers}
import scala.util.parsing.input.CharSequenceReader

// ---------- Exception ----------
final case class ParseException(msg: String, line: Int, column: Int)
  extends RuntimeException(s"Parse error at $line:$column: $msg")

// ---------- Parser ----------
object Parser extends RegexParsers with PackratParsers {

  // Skip whitespace and //-comments
  // (?s) enables dotall to consume newlines in \s
  override val whiteSpace: Regex = """(?s)(?:\s+|//[^\n]*\n?)+""".r

  // -------- Lexical --------
  private val keywords: Set[String] =
    Set("fn", "fix", "if", "then", "else", "true", "false", "Int", "Bool")

  lazy val ident: PackratParser[String] =
    """[A-Za-z][A-Za-z0-9_]*""".r flatMap { s =>
      if keywords.contains(s) then failure(s"identifier expected, got keyword '$s'") else success(s)
    }

  lazy val intLit: PackratParser[Long] =
    """-?\d+""".r ^^ (_.toLong)

  // -------- Types (right-assoc) --------
  lazy val typeExpr: PackratParser[Type] =
    typeAtom ~ opt("->" ~> typeExpr) ^^ {
      case a ~ Some(b) => Type.Arrow(a, b)
      case a ~ None    => a
    }

  lazy val typeAtom: PackratParser[Type] =
    ("Int"  ^^^ Type.IntType) |
      ("Bool" ^^^ Type.BoolType) |
      recordType |
      "(" ~> typeExpr <~ ")"

  lazy val recordType: PackratParser[Type] =
    "{" ~> repsep(ident ~ (":" ~> typeExpr), ",") <~ "}" ^^ { fields =>
      Type.RecordType(fields.map { case name ~ tpe => (name, tpe) }.toMap)
    }

  // -------- Expressions --------
  lazy val program: PackratParser[Expr] =
    phrase(expr)

  lazy val expr: PackratParser[Expr] =
    lambda | fix | ifExpr | comparison

  // \x: T. e   |  fn x: T => e
  lazy val lambda: PackratParser[Expr] =
    (("\\" | "fn") ~> ident) ~ (":" ~> typeExpr) ~ (("." | "=>") ~> expr) ^^ {
      case name ~ tpe ~ body => Expr.Lam(name, tpe, body)
    }

  // fix f: T. e
  lazy val fix: PackratParser[Expr] =
    ("fix" ~> ident) ~ (":" ~> typeExpr) ~ ("." ~> expr) ^^ {
      case name ~ tpe ~ body => Expr.Fix(name, tpe, body)
    }

  // if e1 then e2 else e3
  lazy val ifExpr: PackratParser[Expr] =
    ("if" ~> expr) ~ ("then" ~> expr) ~ ("else" ~> expr) ^^ {
      case c ~ t ~ e => Expr.If(c, t, e)
    }

  // comparison has lower precedence than arithmetic
  lazy val comparison: PackratParser[Expr] =
    arith ~ rep(compOp ~ arith) ^^ {
      case first ~ rest => rest.foldLeft(first) {
        case (acc, "==" ~ rhs) => Expr.BinOpCmp(CmpOpKind.Eq, acc, rhs)
        case (acc, "<"  ~ rhs) => Expr.BinOpCmp(CmpOpKind.Lt, acc, rhs)
        case (acc, ">"  ~ rhs) => Expr.BinOpCmp(CmpOpKind.Gt, acc, rhs)
        case (_, _ ~ _) => throw IllegalStateException("unreachable")
      }
    }

  lazy val compOp: PackratParser[String] =
    "==" | "<" | ">"

  // addition/subtraction
  lazy val arith: PackratParser[Expr] =
    term ~ rep(addOp ~ term) ^^ {
      case first ~ rest =>
        rest.foldLeft(first) {
          case (acc, "+" ~ rhs) => Expr.BinOpInt(IntOpKind.Add, acc, rhs)
          case (acc, "-" ~ rhs) => Expr.BinOpInt(IntOpKind.Sub, acc, rhs)
          case (_, _ ~ _) => throw IllegalStateException("unreachable")
        }
    }

  lazy val addOp: PackratParser[String] =
    "+" | "-"

  // multiplication
  lazy val term: PackratParser[Expr] =
    factor ~ rep("*" ~> factor) ^^ {
      case first ~ rest => rest.foldLeft(first)((acc, rhs) => Expr.BinOpInt(IntOpKind.Mul, acc, rhs))
    }

  // application: atom atom ... (left-assoc)
  lazy val factor: PackratParser[Expr] =
    atom ~ rep(atom) ^^ {
      case first ~ args => args.foldLeft(first)(Expr.App(_, _))
    }

  lazy val atom: PackratParser[Expr] =
    (intLit ^^ (n => Expr.IntLit(n))) |
    ("true"  ^^^ Expr.BoolLit(true)) |
    ("false" ^^^ Expr.BoolLit(false)) |
    recordLiteral |
    (ident ^^ (s => Expr.Var(s))) |
    ("(" ~> expr <~ ")") |
    selectChain

  lazy val recordLiteral: PackratParser[Expr] =
    "{" ~> repsep(ident ~ ("=" ~> expr), ",") <~ "}" ^^ { fields =>
      Expr.Record(fields.map { case name ~ value => (name, value) }.toMap)
    }

  lazy val selectChain: PackratParser[Expr] =
    (ident | ("(" ~> expr <~ ")")) ~ rep("." ~> ident) ^^ {
      case base ~ Nil => base match {
        case s: String => Expr.Var(s)
        case e: Expr => e
      }
      case base ~ fields =>
        val baseExpr = base match {
          case s: String => Expr.Var(s)
          case e: Expr => e
        }
        fields.foldLeft(baseExpr)((acc, field) => Expr.Proj(acc, field))
    }

  // -------- Public API --------
  def parse(input: String): Expr = {
    val reader = new PackratReader(new CharSequenceReader(input))
    program(reader) match {
      case Success(ast, _) => ast
      case NoSuccess(msg, next) =>
        throw ParseException(msg, next.pos.line, next.pos.column)
      case Failure(msg, next) =>
        throw ParseException(msg, next.pos.line, next.pos.column)
      case Error(msg, next) =>
        throw ParseException(msg, next.pos.line, next.pos.column)
    }
  }
}
